File: server-python/services/curriculum_loader.py
"""
Curriculum context loader for AI question generation.
Loads structured curriculum data from JSON files to provide context for AI-generated questions.
"""
import json
import os

# Path to curriculum JSON files (project root books directory)
# This works whether running from server-python or project root
CURRICULUM_DIR = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 'books')
if not os.path.exists(CURRICULUM_DIR):
    # Fallback: try relative to current working directory
    CURRICULUM_DIR = os.path.join(os.getcwd(), 'books')

def get_curriculum_for_grade(grade):
    """
    Load curriculum JSON for a specific grade level.
    
    Args:
        grade: Grade level (7 or 8)
    
    Returns:
        dict: Curriculum data or None if not found
    """
    filename = os.path.join(CURRICULUM_DIR, f'Grade_{grade}_Maths.json')
    
    if not os.path.exists(filename):
        return None
    
    try:
        with open(filename, 'r', encoding='utf-8') as f:
            return json.load(f)
    except (json.JSONDecodeError, IOError) as e:
        print(f"Error loading curriculum file {filename}: {e}")
        return None

def find_topic_in_curriculum(grade, topic_name=None, chapter=None, subtopic=None):
    """
    Find a specific topic in the curriculum and return its context.
    
    Args:
        grade: Grade level (7 or 8)
        topic_name: Topic name to search for (e.g., "Fractions", "1.1 Divisibility tests")
        chapter: Chapter name or number (e.g., "Chapter 1", "Chapter 1: Factors")
        subtopic: Specific subtopic name (e.g., "1.1 Divisibility tests")
    
    Returns:
        dict: Topic context with learning objectives, key concepts, etc. or None
    """
    curriculum = get_curriculum_for_grade(grade)
    if not curriculum:
        return None
    
    # Normalize search terms
    topic_search = (topic_name or '').lower().strip()
    chapter_search = (chapter or '').lower().strip()
    subtopic_search = (subtopic or '').lower().strip() if isinstance(subtopic, str) else ''
    
    # If subtopic is a list (JSON array), use the first item
    if isinstance(subtopic, list) and len(subtopic) > 0:
        subtopic_search = subtopic[0].lower().strip()
    
    for curr_chapter in curriculum['chapters']:
        chapter_name = curr_chapter.get('chapter', '').lower()
        
        # Check if chapter matches
        chapter_matches = (
            chapter_search and (
                chapter_search in chapter_name or 
                chapter_name.startswith(chapter_search) or
                chapter_search.startswith(chapter_name.split(':')[0])
            )
        ) or (
            topic_search and topic_search in chapter_name
        )
        
        # Search through topics in this chapter
        for topic in curr_chapter.get('topics', []):
            topic_name_field = topic.get('name', '').lower()
            
            # Check if topic matches
            topic_matches = (
                topic_search and topic_search in topic_name_field
            ) or (
                subtopic_search and (
                    subtopic_search in topic_name_field or
                    topic_name_field.startswith(subtopic_search)
                )
            ) or (
                chapter_matches and True  # If chapter matches, return first topic
            )
            
            if topic_matches:
                return {
                    'topic': topic,
                    'chapter': curr_chapter,
                    'grade': grade,
                    'curriculum_metadata': {
                        'book': curriculum.get('book', ''),
                        'grade_level': curriculum.get('grade_level', grade),
                        'curriculum': curriculum.get('curriculum', ''),
                    }
                }
    
    # No specific topic found, return general curriculum info
    return {
        'topic': None,
        'chapter': None,
        'grade': grade,
        'curriculum_metadata': {
            'book': curriculum.get('book', ''),
            'grade_level': curriculum.get('grade_level', grade),
            'curriculum': curriculum.get('curriculum', ''),
        }
    }

def build_curriculum_context(grade, topic_name=None, chapter=None, subtopic=None):
    """
    Build a formatted context string for AI prompts.
    
    Args:
        grade: Grade level (7 or 8)
        topic_name: Topic name
        chapter: Chapter name
        subtopic: Subtopic name
    
    Returns:
        str: Formatted curriculum context for AI prompts
    """
    result = find_topic_in_curriculum(grade, topic_name, chapter, subtopic)
    
    if not result or not result['topic']:
        # Return minimal context if topic not found
        return f"""CURRICULUM CONTEXT:
Grade {grade} Mathematics
Note: Specific topic not found in curriculum. Generate age-appropriate questions."""
    
    topic = result['topic']
    chapter_data = result['chapter']
    metadata = result['curriculum_metadata']
    
    # Build formatted context
    context_parts = [
        "CURRICULUM CONTEXT:",
        f"Textbook: {metadata.get('book', 'Unknown')}",
        f"Grade Level: {metadata.get('grade_level', grade)}",
        f"Chapter: {chapter_data.get('chapter', 'Unknown')}",
        f"Topic: {topic.get('name', 'Unknown')}",
        f"Book Pages: {topic.get('book1_pages', 'N/A')}",
        "",
        "LEARNING OBJECTIVES:",
    ]
    
    for obj in topic.get('learning_objectives', []):
        context_parts.append(f"  - {obj}")
    
    context_parts.append("")
    context_parts.append("KEY CONCEPTS:")
    for concept in topic.get('key_concepts', []):
        context_parts.append(f"  - {concept}")
    
    context_parts.append("")
    context_parts.append("TERMINOLOGY (use these exact terms):")
    for term in topic.get('terminology', []):
        context_parts.append(f"  - {term}")
    
    context_parts.append("")
    context_parts.append("SUBSKILLS TO TEST:")
    for skill in topic.get('subskills', []):
        context_parts.append(f"  - {skill}")
    
    example_problems = topic.get('example_problems', [])
    if example_problems:
        context_parts.append("")
        context_parts.append("EXAMPLE PROBLEM STYLES (match these formats):")
        for prob in example_problems[:3]:  # Limit to 3 examples
            context_parts.append(f"  - {prob}")
    
    common_mistakes = topic.get('common_mistakes', [])
    if common_mistakes:
        context_parts.append("")
        context_parts.append("COMMON MISTAKES (use as distractors in multiple choice):")
        for mistake in common_mistakes:
            context_parts.append(f"  - {mistake}")
    
    question_style = topic.get('question_style_notes', '')
    if question_style:
        context_parts.append("")
        context_parts.append("QUESTION STYLE GUIDANCE:")
        context_parts.append(f"  {question_style}")
    
    real_life_apps = topic.get('real_life_applications', [])
    if real_life_apps:
        context_parts.append("")
        context_parts.append("REAL-LIFE APPLICATIONS (for word problems):")
        for app in real_life_apps[:2]:  # Limit to 2
            context_parts.append(f"  - {app}")
    
    return "\n".join(context_parts)

def get_all_topics_for_grade(grade):
    """
    Get all topics for a grade level as a flat list.
    
    Args:
        grade: Grade level (7 or 8)
    
    Returns:
        list: List of topic dictionaries with chapter info
    """
    curriculum = get_curriculum_for_grade(grade)
    if not curriculum:
        return []
    
    topics = []
    for chapter in curriculum['chapters']:
        for topic in chapter.get('topics', []):
            topics.append({
                'chapter': chapter.get('chapter', ''),
                'topic': topic,
                'grade': grade
            })
    
    return topics

# Test function
if __name__ == '__main__':
    # Test loading Grade 7 curriculum
    print("Testing Grade 7 curriculum loader...")
    context = build_curriculum_context(7, "Divisibility", "Chapter 1")
    print(context)
    print("\n" + "="*80 + "\n")
    
    # Test loading Grade 8 curriculum
    print("Testing Grade 8 curriculum loader...")
    context = build_curriculum_context(8, "Negative numbers", "Chapter 1")
    print(context)

File: client/src/utils/curriculumLoader.js
/**
 * Curriculum context loader for frontend AI question generation.
 * Loads structured curriculum data from JSON files to provide context for AI-generated questions.
 */

// Curriculum data will be loaded dynamically since books/ is outside the src/ directory
let grade7Curriculum = null;
let grade8Curriculum = null;

/**
 * Load curriculum data dynamically.
 * Note: books/ directory is at project root, not bundled with Vite.
 * We'll fetch from the public path or use embedded data.
 */
async function loadCurriculumData(grade) {
  const cache = grade === 7 ? grade7Curriculum : grade8Curriculum;
  if (cache) return cache;
  
  // Parse grade to ensure it's a number
  const gradeStr = String(grade || '');
  const gradeNum = parseInt(gradeStr.replace(/[^0-9]/g, ''), 10);
  
  console.log('[curriculumLoader] loadCurriculumData called with:', { 
    original: grade, 
    parsed: gradeNum,
    gradeStr 
  });
  
  if (!gradeNum || (gradeNum !== 7 && gradeNum !== 8)) {
    console.warn(`Invalid grade: ${grade} (parsed: ${gradeNum}), using grade 7 as fallback`);
    return null;
  }
  
  try {
    // Try to fetch from the server (books should be served statically or via API)
    const response = await fetch(`/books/Grade_${gradeNum}_Maths.json`);
    if (!response.ok) throw new Error('Failed to load curriculum');
    const data = await response.json();
    
    if (gradeNum === 7) {
      grade7Curriculum = data;
    } else {
      grade8Curriculum = data;
    }
    
    console.log('[curriculumLoader] Successfully loaded Grade', gradeNum, 'curriculum');
    return data;
  } catch (error) {
    console.warn(`Could not load Grade ${gradeNum} curriculum, using fallback:`, error.message);
    return null;
  }
}

/**
 * Get curriculum data for a specific grade level.
 * @param {number|string} grade - Grade level (7 or 8)
 * @returns {object|null} Curriculum data or null
 */
export async function getCurriculumForGrade(grade) {
  // Handle "Grade 7", "7", 7, "grade7", etc.
  const gradeNum = typeof grade === 'number' 
    ? grade 
    : parseInt(String(grade || '').replace(/[^0-9]/g, ''), 10);
  return await loadCurriculumData(gradeNum);
}

/**
 * Find a specific topic in the curriculum.
 * @param {number|string} grade - Grade level
 * @param {string} topicName - Topic name to search for
 * @param {string} chapter - Chapter name
 * @param {string|array} subtopic - Subtopic name or array
 * @returns {Promise<object|null>} Topic context or null
 */
export async function findTopicInCurriculum(grade, topicName = null, chapter = null, subtopic = null) {
  const gradeNum = parseInt(String(grade || '').replace(/[^0-9]/g, ''), 10) || 7;
  const curriculum = await getCurriculumForGrade(gradeNum);
  if (!curriculum) return null;

  const topicSearch = (topicName || '').toLowerCase().trim();
  const chapterSearch = (chapter || '').toLowerCase().trim();
  const subtopicSearch = Array.isArray(subtopic) 
    ? (subtopic[0] || '').toLowerCase().trim()
    : (subtopic || '').toLowerCase().trim();

  for (const currChapter of curriculum.chapters) {
    const chapterName = (currChapter.chapter || '').toLowerCase();

    for (const topic of currChapter.topics || []) {
      const topicNameField = (topic.name || '').toLowerCase();

      // Check if chapter matches (by chapter number or name)
      const chapterMatches = chapterSearch && (
        chapterName.includes(chapterSearch) ||
        chapterSearch.includes(chapterName.split(':')[0]) ||
        topicNameField.includes(chapterSearch)
      );

      // Check if topic matches
      const topicMatches = (
        topicSearch && (
          topicNameField.includes(topicSearch) ||
          topicSearch.includes(topicNameField.split(' ')[0])
        )
      ) || (
        subtopicSearch && (
          topicNameField.includes(subtopicSearch) ||
          topicNameField.startsWith(subtopicSearch)
        )
      );

      if (topicMatches || chapterMatches) {
        return {
          topic,
          chapter: currChapter,
          grade: gradeNum,
          curriculumMetadata: {
            book: curriculum.book || '',
            gradeLevel: curriculum.grade_level || gradeNum,
            curriculum: curriculum.curriculum || '',
          },
        };
      }
    }
  }

  return null;
}

/**
 * Build a formatted context string for AI prompts.
 * @param {number|string} grade - Grade level
 * @param {string} topicName - Topic name
 * @param {string} chapter - Chapter name
 * @param {string|array} subtopic - Subtopic name or array
 * @returns {Promise<string>} Formatted curriculum context
 */
export async function buildCurriculumContext(grade, topicName = null, chapter = null, subtopic = null) {
  const gradeNum = parseInt(String(grade || '').replace(/[^0-9]/g, ''), 10) || 7;
  const result = await findTopicInCurriculum(gradeNum, topicName, chapter, subtopic);

  if (!result || !result.topic) {
    return `CURRICULUM CONTEXT:
Grade ${gradeNum} Mathematics
Note: Specific topic not found in curriculum. Generate age-appropriate questions.`;
  }

  const { topic, chapter: chapterData, curriculumMetadata } = result;
  const lines = [
    'CURRICULUM CONTEXT:',
    `Textbook: ${curriculumMetadata.book || 'Unknown'}`,
    `Grade Level: ${curriculumMetadata.gradeLevel || grade}`,
    `Chapter: ${chapterData.chapter || 'Unknown'}`,
    `Topic: ${topic.name || 'Unknown'}`,
    `Book Pages: ${topic.book1_pages || 'N/A'}`,
    '',
    'LEARNING OBJECTIVES:',
  ];

  (topic.learning_objectives || []).forEach(obj => {
    lines.push(`  - ${obj}`);
  });

  lines.push('');
  lines.push('KEY CONCEPTS:');
  (topic.key_concepts || []).forEach(concept => {
    lines.push(`  - ${concept}`);
  });

  lines.push('');
  lines.push('TERMINOLOGY (use these exact terms):');
  (topic.terminology || []).forEach(term => {
    lines.push(`  - ${term}`);
  });

  lines.push('');
  lines.push('SUBSKILLS TO TEST:');
  (topic.subskills || []).forEach(skill => {
    lines.push(`  - ${skill}`);
  });

  if (topic.example_problems?.length) {
    lines.push('');
    lines.push('EXAMPLE PROBLEM STYLES (match these formats):');
    topic.example_problems.slice(0, 3).forEach(prob => {
      lines.push(`  - ${prob}`);
    });
  }

  if (topic.common_mistakes?.length) {
    lines.push('');
    lines.push('COMMON MISTAKES (use as distractors in multiple choice):');
    topic.common_mistakes.forEach(mistake => {
      lines.push(`  - ${mistake}`);
    });
  }

  if (topic.question_style_notes) {
    lines.push('');
    lines.push('QUESTION STYLE GUIDANCE:');
    lines.push(`  ${topic.question_style_notes}`);
  }

  if (topic.real_life_applications?.length) {
    lines.push('');
    lines.push('REAL-LIFE APPLICATIONS (for word problems):');
    topic.real_life_applications.slice(0, 2).forEach(app => {
      lines.push(`  - ${app}`);
    });
  }

  return lines.join('\n');
}

/**
 * Get all topics for a grade level as a flat list.
 * @param {number|string} grade - Grade level
 * @returns {Promise<array>} List of topic dictionaries
 */
export async function getAllTopicsForGrade(grade) {
  const curriculum = await getCurriculumForGrade(grade);
  if (!curriculum) return [];

  const topics = [];
  curriculum.chapters.forEach(chapter => {
    (chapter.topics || []).forEach(topic => {
      topics.push({
        chapter: chapter.chapter || '',
        topic,
        grade: Number(grade),
      });
    });
  });

  return topics;
}
